zcbor.py: Add support for unordered maps in generated code

oyvindronningstad opened the pull request:
Add support for the --unordered-maps option to zcbor code
which makes use of the zcbor_unordered_maps_*() API in generated code.

Fixes #411

oyvindronningstad committed:
zcbor.py: Add support for unordered maps in generated code

Add support for the --unordered-maps option to zcbor code
which makes use of the zcbor_unordered_maps_*() API in generated code.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

simensrostad commented:
@oyvindronningstad Any traction here? Will we get this for NCS 2.8?

darkofc commented on include/zcbor_common.h:
```diff
@@ -530,7 +530,7 @@ static inline size_t zcbor_flags_to_states(size_t num_flags)
 #define ZCBOR_FLAG_STATES(n_flags) zcbor_flags_to_states(n_flags)
```
The changes to zcbor.py is using ZCBOR_FLAG_STATES() in the states[] array size.  The IAR C Compiler does not allow inline methods in array size declarations.   I was able to work around this by expanding the ZCBOR_FLAG_STATES()  macro and not calling zcbor_flags_to_states() (which is then unused and can be removed) .  Not sure if that is the best solution, but seems to be working:

```
#define ZCBOR_ROUND_UP(x, align) (((x) + (align) - 1) / (align) * (align))
#define ZCBOR_FLAG_STATES(num_flags) (ZCBOR_ROUND_UP(num_flags, sizeof(zcbor_state_t) * ZCBOR_BITS_PER_BYTE) / (sizeof(zcbor_state_t) * ZCBOR_BITS_PER_BYTE))

```

oyvindronningstad replied:
Thanks for the heads up. Fixed in #456.

darkofc commented on zcbor/zcbor.py:
```diff
@@ -3019,7 +3143,7 @@ def relativify(p):
     {(linesep + "    ").join(((str(relativify(f)) for f in include_dirs)))}
     )
-"""
+{f"\n{smart_search}\n" if add_smart_search else ''}"""
```
Nit: The use of backslashes in the formatted string doesn't work on version 3.10 python - forced me to update python to pickup this change.

jtguggedal commented:
@oyvindronningstad @darkofc It would be really great if this PR could be merged so that we can start to use this feature

oyvindronningstad commented:
@jtguggedal I'm working on it. Sorry it's taken a long time, but each time I come back to it, I change it quite a bit, because it would be prone to bugs or user error. It contains surprisingly many tricky edge cases and tough decisions. Hoping to finish it soon.

oyvindronningstad committed:
zcbor.py: Fix bug in quote parsing.

Nested braces and quotes produced a recursion error.
This happened because the quotes regex didn't disallow quotes in between
quotes. I thought this was unneccessary because the *? occurence
specifier would never take characters beyond the quotes. But this
limitation is broken if another constraint, such as looking for commas
is not hit. In that case, the regex engine backtracks and loosens the
*? constraint, taking additional letters. Adding an explicit disallowing
of the quotes seems to fix this problem.

Thanks to @Trond-F-Christiansen for finding, debugging, and providing
test cases for this bug!

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>
